pkg/component/runtime/command.go

```diff
@@ -392,8 +396,18 @@ func (c *CommandRuntime) handleProc(state *os.ProcessState) bool {
 	switch c.actionState {
 	case actionStart:
 		// should still be running
-		stopMsg := fmt.Sprintf("Failed: pid '%d' exited with code '%d'", state.Pid(), state.ExitCode())
-		c.forceCompState(client.UnitStateFailed, stopMsg)
+		reportFailure := true
+		if c.restartBucket != nil {
+			// report failure only if bucket is full of restart events
+			reportFailure = !c.restartBucket.Allow()
+		}
+
+		if reportFailure {
+			stopMsg := fmt.Sprintf("Failed: pid '%d' exited with code '%d'", state.Pid(), state.ExitCode())
+			c.forceCompState(client.UnitStateFailed, stopMsg)
+		} else {
+			c.logger.Info("Suppressing FAILED state due to restart for '%d' exited with code '%d'", state.Pid(), state.ExitCode())
```

Comment:
I think it might be better to transition the unit to stopped here, or per my comment above in https://github.com/elastic/elastic-agent/pull/2003/files#r1058695336 add some logic above that will clear out the current unit states so the agent is in sync with the Beats.

Possibly a modification to or a new variant of this function:

https://github.com/elastic/elastic-agent/blob/668fa22918fa330db979fc9fbd81ef3854deb4c2/pkg/component/runtime/state.go#L377